vault/router.go

```diff
@@ -205,6 +212,39 @@ func (r *Router) MatchingMount(path string) string {
 	return mount
 }
 
+// matchingPrefixInternal returns a mount prefix that a path may be a part of
+func (r *Router) matchingPrefixInternal(path string) string {
+	var existing string = ""
+	fn := func(existing_path string, _v interface{}) bool {
+		if strings.HasPrefix(existing_path, path) {
+			existing = existing_path
+			return true
+		}
+		return false
+	}
+	r.root.WalkPrefix(path, fn)
+	return existing
+}
+
+// MatchingPrefix is like matchingPrefixInternal but locks the router
+func (r *Router) MatchingPrefix(path string) string {
+	r.l.RLock()
+	match := r.matchingPrefixInternal(path)
+	r.l.RUnlock()
+	return match
+}
+
+// MountConflict determines if there are potential path conflicts
+func (r *Router) MountConflict(path string) string {
+	if exact_match := r.MatchingMount(path); exact_match != "" {
```

Comment:
Really this function should grab a single read lock for the entire function and then call matchingMountInternal and matchingPrefixInternal. That way nothing can change between the two functions.